server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/RequestHandler.java

```diff
@@ -77,24 +77,24 @@ public RequestHandler(Set<Method> methods, JmapRequestParser jmapRequestParser,
                         return method.process(jmapRequest, request.getClientId(), mailboxSession);
                     } catch (IOException e) {
                         LOGGER.error("Error occured while parsing the request.", e);
-                        if (e.getCause() instanceof NotImplementedException) {
-                            return errorNotImplemented(request);
+                        if (e.getCause() instanceof JmapNotImplementedException) {
+                            return errorNotImplemented((JmapNotImplementedException) e.getCause(), request);
                         }
                         return error(request, ErrorResponse.builder()
                                                         .type("invalidArguments")
                                                         .description(e.getMessage())
                                                         .build());
-                    } catch (NotImplementedException e) {
-                        return errorNotImplemented(request);
+                    } catch (JmapNotImplementedException e) {
+                        return errorNotImplemented(e, request);
                     }
                 };
     }
 
-    private Stream<JmapResponse> errorNotImplemented(AuthenticatedProtocolRequest request) {
+    private Stream<JmapResponse> errorNotImplemented(Exception error, AuthenticatedProtocolRequest request) {
         return Stream.of(
                 JmapResponse.builder()
                     .clientId(request.getClientId())
-                    .error("Not yet implemented")
+                    .error(error.getMessage())
```

Comment:
It's a bit sad.

Thanks to the typed exceptions we can rely on fields to convey the not implemented field & the model in which it was specified... This would allow formatting the error message here, and we would not take the risk to disclose sensible informations through the error message...

Reply:
Reading `ErrorResponse` & https://jmap.io/spec-core.html#errors I think you are mixing `type` and `description`.

The `'accountId' is not yet implemented 'getMessages'` is a description while the type would be `invalidArguments`.

Could you please apply that change?

Reply:
I see the point of your first message, but we have to types of messages, we have to rely on polymorphism to deal with that.

Reply:
Writing 2 catch clauses would seem more straitghforward to me.

Globally, I'd say that the pattern of good software is to have "responsibility concerns" shaping "technical concerns" like class hierarchy. Not to have "technical concerns" shaping the "responsibilities".

Do we need that class hierarchy?

Is having two catch blocks that bad?

Reply:
Well, we don't even need two classes.